File: CSV2HTML_by_Java_Skeleton/csv2html/Writer.java
package csv2html;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.UnsupportedEncodingException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import utility.StringUtility;

/**
 * ライタ：情報のテーブルをHTMLページとして書き出す。
 */
public class Writer extends IO {

	/**
	 * ライタのコンストラクタ。
	 *
	 * @param aTable テーブル
	 */
	public Writer(Table aTable) {
		super(aTable);
		return;
	}

	/**
	 * HTMLページを基にするテーブルからインデックスファイル(index.html)に書き出す。
	 */
	public void perform() {
		try {
			Attributes attributes = this.attributes();
			String fileStringOfHTML = attributes.baseDirectory() + attributes.indexHTML();
			File aFile = new File(fileStringOfHTML);
			FileOutputStream outputStream = new FileOutputStream(aFile);
			OutputStreamWriter outputWriter = new OutputStreamWriter(outputStream, StringUtility.encodingSymbol());
			BufferedWriter aWriter = new BufferedWriter(outputWriter);

			this.writeHeaderOn(aWriter);
			this.writeTableBodyOn(aWriter);
			this.writeFooterOn(aWriter);

			aWriter.close();
		} catch (UnsupportedEncodingException | FileNotFoundException anException) {
			anException.printStackTrace();
		} catch (IOException anException) {
			anException.printStackTrace();
		}

		return;
	}

	/**
	 * ボディを書き出す。
	 *
	 * @param aWriter ライタ
	 */
	public void writeTableBodyOn(BufferedWriter aWriter) {

		try {
			aWriter.write("<body>\n");
			aWriter.write("<div class=\"belt\">\n");
			aWriter.write("<h2>" + this.attributes().captionString() + "</h2>\n");
			aWriter.write("</div>\n");
			aWriter.write("<table class=\"belt\" summary=\"table\">\n");
			aWriter.write("	<tbody>\n");
			aWriter.write("		<tr>\n");
			aWriter.write("			<td>\n");
			aWriter.write("				<table class=\"content\" summary=\"table\">\n");
			aWriter.write("					<tbody>\n");

			this.writeAttributesOn(aWriter);
			this.writeTuplesOn(aWriter);

			aWriter.write("					</tbody>\n");
			aWriter.write("				</table>\n");
			aWriter.write("			</td>\n");
			aWriter.write("		</tr>\n");
			aWriter.write("	</tbody>\n");
			aWriter.write("</table>\n");

		} catch (IOException anException) {
			anException.printStackTrace();
		}
		return;
	}

	/**
	 * 属性リストを書き出す。
	 *
	 * @param aWriter ライタ
	 */
	public void writeAttributesOn(BufferedWriter aWriter) {
		String aTab = "						";
		String aStereotypedPhrase = "	<td class=\"center-pink\"><strong>";
		try {
			aWriter.write(aTab + "<tr>");
			aWriter.newLine();

			for (String aString : this.table().attributes().names()) {
				aWriter.write(aTab + aStereotypedPhrase + aString);
				aWriter.write("</strong></td>");
				aWriter.newLine();
			}
			aWriter.write(aTab + "</tr>");
			aWriter.newLine();
		} catch (IOException anException) {
			anException.printStackTrace();
		}

		return;
	}

	/**
	 * タプル群を書き出す。
	 *
	 * @param aWriter ライタ
	 */
	public void writeTuplesOn(BufferedWriter aWriter) {

		String aTab = "						";
		String aStereotypedPhrase;

		try {

			boolean color = true;
			for (Tuple aTuple : this.table().tuples()) {
				aWriter.write(aTab + "<tr>");
				aWriter.newLine();

				if (color) {
					aStereotypedPhrase = "	<td class=\"center-blue\">";
					color = false;
				} else {
					aStereotypedPhrase = "	<td class=\"center-yellow\">";
					color = true;
				}

				for (String aString : aTuple.values()) {
					aWriter.write(aTab + aStereotypedPhrase + aString);
					aWriter.write("</td>");
					aWriter.newLine();
				}
				aWriter.write(aTab + "</tr>");
				aWriter.newLine();
			}
		} catch (Exception anException) {
			anException.printStackTrace();
		}

		return;
	}

	/**
	 * フッタを書き出す。
	 *
	 * @param aWriter ライタ
	 */
	public void writeFooterOn(BufferedWriter aWriter) {
		SimpleDateFormat aSimpleDateFormat = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss");
		Calendar aCalendar = Calendar.getInstance();
		String footer = """
				<hr>
				<div class="right-small">Created by csv2html.Translator on %s</div>
				</body>
				</html>

				""".formatted(aSimpleDateFormat.format(aCalendar.getTime()));

		try {
			aWriter.write(footer);
		} catch (IOException anException) {
			anException.printStackTrace();
		}
		return;
	}

	/**
	 * ヘッダを書き出す。
	 *
	 * @param aWriter ライタ
	 */
	public void writeHeaderOn(BufferedWriter aWriter) {
		String doctype = """
				<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
				<html lang="ja">
				<head>
				""";
		String meta = """
				<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
				<meta http-equiv="Content-Style-Type" content="text/css">
				<meta http-equiv="Content-Script-Type" content="text/javascript">
				<meta name="keywords" content="Smalltalk,Oriented,Programming">
				<meta name="description" content="Prime Ministers">
				<meta name="author" content="AOKI Atsushi">
				""";
		String link = """
				<link rev="made" href="http://www.cc.kyoto-su.ac.jp/~atsushi/">
				<link rel="index" href="index.html">
				""";
		String css = """
				<style type="text/css">
				<!--
				body {
					background-color : #ffffff;
					margin : 20px;
					padding : 10px;
					font-family : serif;
					font-size : 10pt;
				}
				a {
					text-decoration : underline;
					color : #000000;
				}
				a:link {
					background-color : #ffddbb;
				}
				a:visited {
					background-color : #ccffcc;
				}
				a:hover {
					background-color : #dddddd;
				}
				a:active {
					background-color : #dddddd;
				}
				div.belt {
					background-color : #eeeeee;
					padding : 0px 4px;
				}
				div.right-small {
					text-align : right;
					font-size : 8pt;
				}
				img.borderless {
					border-width : 0px;
					vertical-align : middle;
				}
				table.belt {
					border-style : solid;
					border-width : 0px;
					border-color : #000000;
					background-color : #ffffff;
					padding : 0px 0px;
					width : 100%;
				}
				table.content {
					border-style : solid;
					border-width : 0px;
					border-color : #000000;
					padding : 2px 2px;
				}
				td.center-blue {
					padding : 2px 2px;
					text-align : center;
					background-color : #ddeeff;
				}
				td.center-pink {
					padding : 2px 2px;
					text-align : center;
					background-color : #ffddee;
				}
				td.center-yellow {
					padding : 2px 2px;
					text-align : center;
					background-color : #ffffcc;
				}
				-->
				</style>
				""";
		String title = """
				<title>%s</title>
				</head>
				""".formatted(this.attributes().titleString());

		try {
			aWriter.write(doctype + meta + link + css + title);
		} catch (IOException anException) {
			anException.printStackTrace();
		}
		return;
	}
}

File: CSV2HTML_by_Python_Skeleton/csv2html/writer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""CSV2HTML：総理大臣と徳川幕府の情報「CSV」からWebページ「HTML」を生成。"""
__author__ = 'AOKI Atsushi'
__version__ = '1.0.7'
__date__ = '2021/01/10 (Created: 2016/01/01)'

import datetime
import os
import textwrap

from csv2html.io import IO

# pylint: disable=R0201
# R0201: Method could be a function (no-self-use)

class Writer(IO):
	"""ライタ：情報のテーブルをHTMLページとして書き出す。"""

	def __init__(self, output_table):
		"""ライタのコンストラクタ。HTMLページを基にするテーブルを受け取る。"""

		super().__init__(output_table)
		(lambda x: x)(output_table) # NOP

	def perform(self):
		"""HTMLページを基にするテーブルから、インデックスファイル(index_html)に書き出す。"""

		class_attributes = self.attributes().__class__
		base_directory = class_attributes.base_directory()
		index_html = class_attributes.index_html()

		html_filename = os.path.join(base_directory, index_html)
		with open(html_filename, 'w', encoding='utf-8') as a_file:
			self.write_header(a_file)
			self.write_body(a_file)
			self.write_footer(a_file)

	def write_body(self, file):
		"""ボディを書き出す。つまり、属性リストを書き出し、タプル群を書き出す。"""

		body = textwrap.indent(textwrap.dedent(f"""\
			<body>
				<div class=\"belt\">
					<h2>{self.attributes().caption_string()}</h2>
				</div>
				<table class=\"belt\" summary=\"table\">
					<tbody>
						<tr>
							<td>
								<table class=\"content\" summary=\"table\">
									<tbody>
			"""), '	')

		body += "								<tr>\n"
		# 属性リストを書き出す
		for a_string in self.table().attributes().return_names():
			body += f"									<td class=\"center-pink\"><strong>{a_string}</strong></td>\n"
		body += "								</tr>\n"

		# タプル群を書き出す
		color = True
		for a_tuple in self.table().tuples():
			if(color):
				a_stereotyped_phrase = "									<td class=\"center-blue\">"
				color = False
			else:
				a_stereotyped_phrase = "									<td class=\"center-yellow\">"
				color = True

			body += "								<tr>\n"
			for a_string in a_tuple.values():
				body += a_stereotyped_phrase
				body += a_string
				body += "</td>\n"
			body += "								</tr>\n"

		body += textwrap.indent(textwrap.dedent("""\
							</tbody>
						</table>
					</td>
				</tr>
			</tbody>
		</table>
		"""), '		')

		file.write(body)

	def write_footer(self, file):
		"""フッタを書き出す。"""

		footer = textwrap.dedent(f"""\
				<hr>
				<div class="right-small">Created by csv2html.Translator on {datetime.datetime.now().isoformat(' ', 'seconds')} </div>
			</body>
		</html>
		""")
		file.write(footer)

	def write_header(self, file):
		"""ヘッダを書き出す。"""

		doctype = textwrap.dedent("""\
			<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
			<html lang="ja">
				<head>
			""")
		meta = textwrap.indent(textwrap.dedent("""\
			<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
			<meta http-equiv="Content-Style-Type" content="text/css">
			<meta http-equiv="Content-Script-Type" content="text/javascript">
			<meta name="keywords" content="Smalltalk,Oriented,Programming">
			<meta name="description" content="Prime Ministers">
			<meta name="author" content="AOKI Atsushi">
			"""), '		')
		link = textwrap.indent(textwrap.dedent("""\
			<link rev="made" href="http://www.cc.kyoto-su.ac.jp/~atsushi/">
			<link rel="index" href="index.html">
			"""), '		')
		css = textwrap.indent(textwrap.dedent("""\
			<style type="text/css">
				<!--
				body {
					background-color : #ffffff;
					margin : 20px;
					padding : 10px;
					font-family : serif;
					font-size : 10pt;
				}
				a {
					text-decoration : underline;
					color : #000000;
				}
				a:link {
					background-color : #ffddbb;
				}
				a:visited {
					background-color : #ccffcc;
				}
				a:hover {
					background-color : #dddddd;
				}
				a:active {
					background-color : #dddddd;
				}
				div.belt {
					background-color : #eeeeee;
					padding : 0px 4px;
				}
				div.right-small {
					text-align : right;
					font-size : 8pt;
				}
				img.borderless {
					border-width : 0px;
					vertical-align : middle;
				}
				table.belt {
					border-style : solid;
					border-width : 0px;
					border-color : #000000;
					background-color : #ffffff;
					padding : 0px 0px;
					width : 100%;
				}
				table.content {
					border-style : solid;
					border-width : 0px;
					border-color : #000000;
					padding : 2px 2px;
				}
				td.center-blue {
					padding : 2px 2px;
					text-align : center;
					background-color : #ddeeff;
				}
				td.center-pink {
					padding : 2px 2px;
					text-align : center;
					background-color : #ffddee;
				}
				td.center-yellow {
					padding : 2px 2px;
					text-align : center;
					background-color : #ffffcc;
				}
				-->
			</style>
			"""), '		')
		title = textwrap.indent(textwrap.dedent(f"""\
				<title>
					{self.attributes().title_string()}
				</title>
			</head>
			"""), '	')
		file.write(doctype + meta + link + css + title)
